File: examples/typescript/documentation/variables.ts
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0

// DOCS_BLOCK_START:outputs,outputs-usage,remote-state
import { TerraformLocal, TerraformStack, TerraformVariable } from "cdktf";
import { Construct } from "constructs";
import { App, TerraformOutput } from "cdktf";
// DOCS_BLOCK_END:outputs,outputs-usage,remote-state

// DOCS_BLOCK_START:remote-state
import {
  CloudBackend,
  DataTerraformRemoteState,
  NamedCloudWorkspace,
} from "cdktf";
// DOCS_BLOCK_END:remote-state

// DOCS_BLOCK_START:outputs-usage
import { RandomProvider } from "@cdktf/provider-random/lib/provider";
import { Pet } from "@cdktf/provider-random/lib/pet";
// DOCS_BLOCK_END:outputs-usage

import { Instance } from "@cdktf/provider-aws/lib/instance";
import { AwsProvider } from "@cdktf/provider-aws/lib/aws-provider";

export class VariablesStack extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });

    // DOCS_BLOCK_START:variables
    const imageId = new TerraformVariable(this, "imageId", {
      type: "string",
      default: "ami-abcde123",
      description: "What AMI to use to create an instance",
    });
    new Instance(this, "hello", {
      ami: imageId.value,
      instanceType: "t2.micro",
    });
    // DOCS_BLOCK_END:variables

    // DOCS_BLOCK_START:locals
    const commonTags = new TerraformLocal(this, "common_tags", {
      Service: "service_name",
      Owner: "owner",
    });

    new Instance(this, "example", {
      ami: "ami-abcde123",
      instanceType: "t2.micro",
      tags: commonTags.expression,
    });
    // DOCS_BLOCK_END:locals
  }
}

export function outputVariablesRunner() {
  // DOCS_BLOCK_START:outputs
  interface OutputVariableStackConfig {
    readonly myDomain: string;
  }

  class OutputVariableStack extends TerraformStack {
    constructor(
      scope: Construct,
      name: string,
      config: OutputVariableStackConfig
    ) {
      super(scope, name);

      const { myDomain } = config;

      new TerraformOutput(this, "my-domain", {
        value: myDomain,
      });
    }
  }

  const app = new App();
  new OutputVariableStack(app, "cdktf-producer", {
    myDomain: "example.com",
  });
  app.synth();
  // DOCS_BLOCK_END:outputs
}

export function outputsUsageRunner() {
  // DOCS_BLOCK_START:outputs-usage
  class OutputsUsageStack extends TerraformStack {
    constructor(scope: Construct, name: string) {
      super(scope, name);

      new RandomProvider(this, "random", {});
      const pet = new Pet(this, "pet", {});

      new TerraformOutput(this, "random-pet", {
        value: pet.id,
      });
    }
  }

  const app = new App();
  new OutputsUsageStack(app, "outputs-usage");
  app.synth();
  // DOCS_BLOCK_END:outputs-usage
}

export function remoteStateRunner() {
  // DOCS_BLOCK_START:remote-state

  class Producer extends TerraformStack {
    constructor(scope: Construct, name: string) {
      super(scope, name);

      new CloudBackend(this, {
        organization: "hashicorp",
        workspaces: new NamedCloudWorkspace("producer"),
      });

      new RandomProvider(this, "random", {});
      const pet = new Pet(this, "pet", {});

      new TerraformOutput(this, "random-pet", {
        value: pet.id,
      });
    }
  }

  class Consumer extends TerraformStack {
    constructor(scope: Construct, name: string) {
      super(scope, name);

      new CloudBackend(this, {
        organization: "hashicorp",
        workspaces: new NamedCloudWorkspace("consumer"),
      });

      const remoteState = new DataTerraformRemoteState(this, "remote-pet", {
        organization: "hashicorp",
        workspaces: {
          name: "producer",
        },
      });

      new TerraformOutput(this, "random-remote-pet", {
        value: remoteState.getString("random-pet"),
      });
    }
  }

  const app = new App();
  new Producer(app, "cdktf-producer");
  new Consumer(app, "cdktf-consumer");
  app.synth();
  // DOCS_BLOCK_END:remote-state
}

File: examples/go/documentation/variables.go
package main

import (
	"fmt"

	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
	"github.com/hashicorp/terraform-cdk-go/cdktf"
	"github.com/hashicorp/terraform-cdk/examples/go/documentation/generated/hashicorp/aws/eksnodegroup"
	"github.com/hashicorp/terraform-cdk/examples/go/documentation/generated/hashicorp/aws/instance"
	aws "github.com/hashicorp/terraform-cdk/examples/go/documentation/generated/hashicorp/aws/provider"
)

func NewVariablesStack(scope constructs.Construct, name string) cdktf.TerraformStack {
	stack := cdktf.NewTerraformStack(scope, &name)

	aws.NewAwsProvider(stack, jsii.String("aws"), &aws.AwsProviderConfig{
		Region: jsii.String("eu-central-1"),
	})

	// DOCS_BLOCK_START:variables
	imageId := cdktf.NewTerraformVariable(stack, jsii.String("imageId"), &cdktf.TerraformVariableConfig{
		Type:        jsii.String("string"),
		Default:     jsii.String("ami-abcde123"),
		Description: jsii.String("What AMI to use to create an instance"),
	})
	instance.NewInstance(stack, jsii.String("hello"), &instance.InstanceConfig{
		Ami:          imageId.StringValue(),
		InstanceType: jsii.String("t2.micro"),
	})
	// DOCS_BLOCK_END:variables

	// DOCS_BLOCK_START:variables-complex
	nodeGroupConfig := cdktf.NewTerraformVariable(stack, jsii.String("node-group-config"), &cdktf.TerraformVariableConfig{
		Type: cdktf.VariableType_Object(&map[string]*string{
			"node_group_name": cdktf.VariableType_STRING(),
			"instance_types":  cdktf.VariableType_List(cdktf.VariableType_STRING()),
			"min_size":        cdktf.VariableType_NUMBER(),
			"desired_size":    cdktf.VariableType_NUMBER(),
			"max_size":        cdktf.VariableType_NUMBER(),
		}),
		Nullable:    jsii.Bool(false),
		Description: jsii.String("Node group configuration"),
	})

	eksnodegroup.NewEksNodeGroup(stack, jsii.String("node-group"), &eksnodegroup.EksNodeGroupConfig{
		ClusterName:   jsii.String("my-cluster"),
		NodeRoleArn:   jsii.String("arn:::::dummy"),
		SubnetIds:     jsii.Strings("id-1234"),
		NodeGroupName: jsii.String(fmt.Sprintf("${%s.node_group_name}", *nodeGroupConfig.Fqn())),
		InstanceTypes: cdktf.Token_AsList(fmt.Sprintf("${%s.instance_types}", *nodeGroupConfig.Fqn()), &cdktf.EncodingOptions{}),
		ScalingConfig: &eksnodegroup.EksNodeGroupScalingConfig{
			DesiredSize: cdktf.Token_AsNumber(fmt.Sprintf("${%s.desired_size}", *nodeGroupConfig.Fqn())),
			MinSize:     cdktf.Token_AsNumber(fmt.Sprintf("${%s.min_size}", *nodeGroupConfig.Fqn())),
			MaxSize:     cdktf.Token_AsNumber(fmt.Sprintf("${%s.max_size}", *nodeGroupConfig.Fqn())),
		},
	})
	// DOCS_BLOCK_END:variables-complex

	// DOCS_BLOCK_START:locals
	commonTags := cdktf.NewTerraformLocal(stack, jsii.String("common_tags"), map[string]string{
		"Service": "service_name",
		"Owner":   "owner",
	})
	instance.NewInstance(stack, jsii.String("example"), &instance.InstanceConfig{
		Ami:          jsii.String("ami-abcde123"),
		InstanceType: jsii.String("t2.micro"),
		Tags:         commonTags.AsStringMap(),
	})
	// DOCS_BLOCK_END:locals

	return stack
}
